### PR TITLE
feat(codedeploy): support multiple load balancers for deployment groups

### LINKED ISSUE #27407: (module name): (short issue description)
### Describe the feature

Code Deploy deployment-group.ts only supports a single loadbalancer

https://github.com/aws/aws-cdk/blob/ae24fbff8133c452f72f5c276997e2fb707e054c/packages/aws-cdk-lib/aws-codedeploy/lib/server/deployment-group.ts#L372C11-L372C27

I belive that more then one NLB is supported 

### Use Case

Have more then one NLB in the Deployment group

### Proposed Solution

Change the structrue to take a list of NLBs

### Other Information

_No response_

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

v2.99.1

### Environment details (OS name and version, etc.)

AL2

### PR DESCRIPTION
**Pull request by ghost**
### Summary
This change adds multiple load balancer support for CodeDeploy deployment groups.

There is a currently a loadBalancer field in the ServerDeploymentGroup, adding loadBalancers might be a good field to extend the functionality to create a deployment group with multiple load balancers.

Closes #27407

### CODE REVIEW

**Comment by nicolastremblay1** [comment]
Initial PR can be found [here](https://github.com/aws/aws-cdk/pull/27474), creating a new PR from new feature branch.

File: packages/aws-cdk-lib/aws-codedeploy/test/server/deployment-group.test.ts [hunk 1]
```diff
@@ -212,6 +213,197 @@ describe('CodeDeploy Server Deployment Group', () => {
...
+              'Ref': 'CLBA83A883E',
+            },
+          },
+        ],
+      },
+      'DeploymentStyle': {
+        'DeploymentOption': 'WITH_TRAFFIC_CONTROL',
+      },
+    });
+  });
```
Comment (lpizzinidev) [inline packages/aws-cdk-lib/aws-codedeploy/test/server/deployment-group.test.ts:395]:
```suggestion
  test.each([
    ['CLB + ALB', true, true, false],
    ['CLB + NLB', true, false, true],
    ['ALB + NLB', false, true, true],
    ['CLB + ALB + NLB', true, true, true],
  ])('can be created with multiple load balancers (%s)', (_, withCLB, withALB, withNLB) => {
    const stack = new cdk.Stack();
    const defaultVpc = new ec2.Vpc(stack, 'VPC');

    const loadBalancers: codedeploy.LoadBalancer[] = [];
    const loadBalancerInfo: { TargetGroupInfoList?: any[]; ElbInfoList?: any[] } = {};

    if (withCLB) {
      loadBalancers.push(codedeploy.LoadBalancer.classic(new lbv1.LoadBalancer(stack, `CLB-${i}`, {
        vpc: defaultVpc,
        listeners: [{
          externalPort: 8080,
          externalProtocol: lbv1.LoadBalancingProtocol.TCP,
          internalProtocol: lbv1.LoadBalancingProtocol.TCP,
        }],
      })));

      loadBalancerInfo.ElbInfoList?.push({
        'Name': {
          'Ref': 'CLB677386E0',
        },
      });
    }

    if (withALB) {
      const alb = new lbv2.ApplicationLoadBalancer(stack, 'ALB', {
        vpc: defaultVpc,
      });

      const listener = alb.addListener('ListenerALB', { protocol: lbv2.ApplicationProtocol.HTTP, port: 8080 });
      const targetGroup = listener.addTargets('Fleet', { protocol: lbv2.ApplicationProtocol.HTTP, port: 8080 });
      loadBalancers.push(codedeploy.LoadBalancer.application(targetGroup));

      loadBalancerInfo.TargetGroupInfoList?.push({
        'Name': {
          'Fn::GetAtt': [
            'ALBListenerALBFleetGroupF4CBAA91',
            'TargetGroupName',
          ],
        },
      });
    }

    if (withNLB) {
      const nlb = new lbv2.NetworkLoadBalancer(stack, 'NLB', {
        vpc: defaultVpc,
      });

      const nlbListener = nlb.addListener('ListenerNLB', { port: 8080, protocol: lbv2.Protocol.TCP });
      const nlbTargetGroup = nlbListener.addTargets('Fleet', { port: 8080, protocol: lbv2.Protocol.TCP });
      loadBalancers.push(codedeploy.LoadBalancer.network(nlbTargetGroup));

      loadBalancerInfo.TargetGroupInfoList?.push({
        'Name': {
          'Fn::GetAtt': [
            'NLBListenerNLB0FleetGroup657E42E4',
            'TargetGroupName',
          ],
        },
      });
    }

    new codedeploy.ServerDeploymentGroup(stack, 'DeploymentGroup', {
      loadBalancers,
    });

    Template.fromStack(stack).hasResourceProperties('AWS::CodeDeploy::DeploymentGroup', {
      'LoadBalancerInfo': loadBalancerInfo,
      'DeploymentStyle': {
        'DeploymentOption': 'WITH_TRAFFIC_CONTROL',
      },
    });
  });
```
Thanks for the effort of considering all possible cases and producing a comprehensive test case. 💪
This is my suggestion to keep this more concise and maintainable while testing for all possible combinations of multiple LBs.
Feel free to adjust it or suggest improvements.

File: packages/aws-cdk-lib/aws-codedeploy/README.md [hunk 2]
```diff
@@ -150,6 +150,29 @@ const deploymentGroup = new codedeploy.ServerDeploymentGroup(this, 'DeploymentGr
...
+declare const alb: elb2.ApplicationLoadBalancer;
+const listener = alb.addListener('Listener', { port: 80 });
+const targetGroup = listener.addTargets('Fleet', { port: 80 });
+
+const deploymentGroup = new codedeploy.ServerDeploymentGroup(this, 'DeploymentGroup', {
+  loadBalancers: [
+    codedeploy.LoadBalancer.classic(lb),
+    codedeploy.LoadBalancer.application(targetGroup),
+  ]
+});
```
Comment (lpizzinidev) [inline packages/aws-cdk-lib/aws-codedeploy/README.md:173]:
```suggestion
import * as elb from 'aws-cdk-lib/aws-elasticloadbalancing';
import * as elb2 from 'aws-cdk-lib/aws-elasticloadbalancingv2';

declare const clb: elb.LoadBalancer;
declare const alb: elb2.ApplicationLoadBalancer;
declare const nlb: elb2.NetworkLoadBalancer;

const albListener = alb.addListener('ALBListener', { port: 80 });
const albTargetGroup = albListener.addTargets('ALBFleet', { port: 80 });

const nlbListener = nlb.addListener('NLBListener', { port: 80 });
const nlbTargetGroup = nlbListener.addTargets('NLBFleet', { port: 80 });

const deploymentGroup = new codedeploy.ServerDeploymentGroup(this, 'DeploymentGroup', {
  loadBalancers: [
    codedeploy.LoadBalancer.classic(clb),
    codedeploy.LoadBalancer.application(albTargetGroup),
    codedeploy.LoadBalancer.network(nlbTargetGroup),
  ]
});
```
More complete example.

**Review by lpizzinidev — CHANGES_REQUESTED** [review]
Thanks for opening a new PR and implementing the changes from the last review 👍
I added some suggestions for the test case and documentation.

File: packages/aws-cdk-lib/aws-codedeploy/test/server/deployment-group.test.ts (hunk 1 above)
Comment (nicolastremblay1) [reply packages/aws-cdk-lib/aws-codedeploy/test/server/deployment-group.test.ts:395]: Yes, this is a lot more concise! Thank you for taking a look!

Should we add the case where there could me more than one of the same type of LB?

Comment (lpizzinidev) [reply packages/aws-cdk-lib/aws-codedeploy/test/server/deployment-group.test.ts:395]: I think that with this implementation we're already covering all scenarios.
Adding LBs of the same type vs adding LBs of different types should not make a difference for the scope of this test.

**Review by lpizzinidev — APPROVED** [review]

File: packages/aws-cdk-lib/aws-codedeploy/test/server/deployment-group.test.ts (hunk 1 above)
Comment (nicolastremblay1) [reply packages/aws-cdk-lib/aws-codedeploy/test/server/deployment-group.test.ts]: That's a faire point, all the paths are still covered in your suggested change, it's also cleaner. Added the modification

**Comment by lpizzinidev** [comment]
@nicolastremblay1 
One last note, the title should be changed to [describe the feature](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md#step-4-pull-request).
For example `feat(codedeploy): support multiple load balancers for deployment groups`.

**Review by SankyRed — DISMISSED** [review]
This looks good to me. Thank you for your contribution!

**Pull request merged**